docs: blog for the best practices of goreleaser and cgo (#441)
bytebase/bytebase.com: content/blog/how-we-explored-the-best-practices-of-goreleaser-x-cgo.md

---

> title: How We Explored the Best Practices of GoReleaser X CGO

Reviewer: Title: How to cross compile with CGO using GoReleaser and GitHub Action

filename: how-to-cross-compile-with-cgo-use-goreleaser-and-github-action.md

For SEO

author: Junyi
published_at: 2022/08/24 15:17
feature_image: /static/blog/how-we-explored-the-best-practices-of-goreleaser-x-cgo/banner.webp
tags: Education
description: We meet some trouble using goreleaser and cgo at the same time. This blog tells how we found the best practices. 
---

## Background

When implementing [SQL Review for PostgreSQL](https://www.bytebase.com/docs/sql-review/review-rules/supported-rules), we introduced [pg_query_go](https://github.com/pganalyze/pg_query_go) as the PostgreSQL parser. The pg_query_go uses the native PostgreSQL parser via C bind, which naturally requires CGO support.

> In the [1.2.1 release](https://github.com/Bytebase/Bytebase/tree/release/1.2.1), we found that the [GoReleaser](https://github.com/goreleaser/goreleaser) did not work correctly, and the error message pointed to pg_query_go. Fortunately, this version did not use pg_query_go. So we first added the "!release" Golang tag to ensure a successful release. Then we started the long road to fight against GoReleaser and CGO.

Author: Do you mean chanelog/bytebase-1-2-1?

So why do we need this here? You may have noticed that we have specified four target platforms in the GoReleaser configuration file.
![target_platforms](/static/blog/how-we-explored-the-best-practices-of-goreleaser-x-cgo/target_platforms.webp)

Our build environment is Ubuntu x64 on GitHub Action. Before turning on CGO, we only need to handle the parameters for cross-platform Go compilation. And this step is handled by the Go compiler and GoReleaser. But after CGO is introduced, we also need to compile C/C++ code, so we need the corresponding C/C++ cross-compilation toolchain.

In other words, we may need four compilation toolchains for the target platform. Fortunately,  goreleaser-cross supports them all.
![goreleaser_cross_supported_platforms](/static/blog/how-we-explored-the-best-practices-of-goreleaser-x-cgo/goreleaser_cross_supported_platforms.webp)

Now it looks like the problem is solved, but it is not perfect. The goreleaser/goreleaser-cross repo has only 26 stars, while goreleaser/goreleaser has 10k+ stars. It could mean that using goeleaser-cross have some risks.
![star_history](/static/blog/how-we-explored-the-best-practices-of-goreleaser-x-cgo/star_history.webp)

> If the cross-compilation toolchain is well maintained, the risk is relatively small. The main problem is when the target platform is Darwin. Anyone who has tried to cross-compile from Linux to Darwin knows this is a challenging task. The difficulty is that there is no readily available, well-maintained cross-compilation toolchain. You often need to build your cross-compilation toolchain. It's had to verify the compatibility of the cross-compilation toolchain, and some of them are not well-maintained. The o64-clang/oa64-clang cross-compilation toolchain used by the GoReleaser-cross is built based on the [osxcross](https://github.com/tpoechtrager/osxcross).

Reviewer: `The o64-clang/oa64-clang cross-compilation toolchain used by the GoReleaser-cross is built based on the [osxcross](https://github.com/tpoechtrager/osxcross)`

Not sure what's our intention here, do we suggest that osxcross is a well-maintained one (there is 2.2k star) . But if it's a well-maintained one, then this seems to be a counter-example, because we just mentioned the difficulty of finding a well-maintained toolchain.

Author: Got it. I'll remove it.

![github_action_platforms](/static/blog/how-we-explored-the-best-practices-of-goreleaser-x-cgo/github_action_platforms.webp)

Note that the macOS supported are all x64 architecture. We still need to compile across the architectures but no need to cross the platforms!

So the next step becomes: to compile the Linux binary for both architectures on the Linux platform and compile the Darwin binary for both architectures on the macOS platform.

For the Linux platform

- Compiling from x64 to x64 requires a gcc/g++ toolchain, which we usually use.
- Compiling from x64 to arm64 requires a well-maintained toolchain, "aarch64-linux-gnu-gcc". You can get it directly from the Ubuntu package manager:
```shell
sudo apt-get -y install gcc-aarch64-linux-gnu
```

For the Darwin platform, it's even easier. The clang supports cross-architecture natively. Thanks to LLVM!

After some work, the GitHub Action configuration looks like this.
![github_workflow](/static/blog/how-we-explored-the-best-practices-of-goreleaser-x-cgo/github_workflow.webp)

Two jobs are used to compile binaries on different platforms.

> For the Darwin platform GoReleaser configuration file, we just need to open CGO.

Reviewer: ```suggestion
For the Darwin platform GoReleaser configuration file, we just need to enable CGO.
```

![goreleaser_configuration_darwin](/static/blog/how-we-explored-the-best-practices-of-goreleaser-x-cgo/goreleaser_configuration_darwin.webp)

For the Linux platform GoReleaser configuration file, it's a bit more complicated, and we need to use the corresponding compilation toolchain for the corresponding architecture.
![goreleaser_configuration_linux](/static/blog/how-we-explored-the-best-practices-of-goreleaser-x-cgo/goreleaser_configuration_linux.webp)

We used GoReleaser's ["overrides"](https://goreleaser.com/customization/build/) feature here.

> Up to this point, we can use GoReleaser to compile when CGO is on, but the story is not over yet. The ultimate goal of GoReleaser is to release, and if we use GoReleaser twice as we currently do, we will have two release messages. But this step is easier than before.

Reviewer: ```suggestion
Up to this point, we can use GoReleaser to compile when CGO is on, but the story is not over yet. The ultimate goal of GoReleaser is to release, and if we use GoReleaser twice as we currently do, we will generate two releases. But compared to the cross-compilation headache, this is a piece of cake.
```

## Release!

> The main idea here is simple: separate the building and the release.

Reviewer: ```suggestion
The main idea here is simple: separate the building process from the release.
```

> - In job one, build Linux binaries.
> - In job two, build macOS binaries.
> - Job three depends on job one and job two. Then it does the final release.

Reviewer: ```suggestion
- Job one builds all Linux binaries.
- Job two builds all macOS binaries.
- Job three packages the binaries from job one and two into the final release.
```

> Job one and two skip the release with the skip-publish argument. Job three skips the building and uses the files generated in jobs one and two with the extra_files setting.

Reviewer: ```suggestion
Job one and two skip the release with the skip-publish argument. Job three skips the building stage and uses the files generated in jobs one and two with the extra_files setting.
```

🚀End of story!
![end_of_story](/static/blog/how-we-explored-the-best-practices-of-goreleaser-x-cgo/end_of_story.webp)

## A fun fact

Bytebase already depended on [go-sqlite3](https://github.com/mattn/go-sqlite3), which is also a CGO-dependent package. So why didn't it cause problems before?

> The thing is CGO was not opened in GoReleaser. And go-sqlite3 did a [mock](https://github.com/mattn/go-sqlite3/blob/master/static_mock.go) when CGO was not opened, which means that the go-sqlite3 compiled without CGO was an empty package.

Reviewer: ```suggestion
The thing is CGO was not enabled in GoReleaser. And go-sqlite3 did a [mock](https://github.com/mattn/go-sqlite3/blob/master/static_mock.go) when CGO was not enabled, which means that the go-sqlite3 compiled without CGO was an empty package.
```

![go_sqlite3_mock](/static/blog/how-we-explored-the-best-practices-of-goreleaser-x-cgo/go_sqlite3_mock.webp)

The reason why we didn't encounter problems is that Bytebase had already [migrated from SQLite to PostgreSQL](https://www.bytebase.com/blog/database-migration-sqlite-to-postgresql) before GoReleaser was introduced. So it really didn't use go-sqlite3 and GoReleaser at the same time.

## Summary

> - Don't try to introduce a cross-compilation toolchain unless you absolutely have to. It will incur additional verification and maintenance costs.

Reviewer: I think it worth mentioning GitHub action simplifies a lot because it provides all arch for a platform
